tests/test_ramp_fitting_gls_fit.py

```diff
@@ -0,0 +1,922 @@
+    all zeros.
+    """
+    nints, ngroups, nrows, ncols = 1, 5, 103, 102
+    dims = (nints, ngroups, nrows, ncols)
+    rnoise, gain = 10, 1
+    group_time, frame_time = 1.0, 1
+
+    ramp_data, gain2d, rnoise2d = setup_inputs(
+        dims, gain, rnoise, group_time, frame_time
+    )
+    ramp_data.data[0:, 0:12, :] = np.nan
+
+    save_opt, algo, ncores = False, "GLS", "none"
+    slopes, cube, ols_opt, gls_opt = ramp_fit_data(
+        ramp_data, 512, save_opt, rnoise2d, gain2d, algo,
+        "optimal", ncores, test_dq_flags)
+
+    assert 0 == np.max(slopes[0])
+    assert 0 == np.min(slopes[0])
+
+
+def test_error_when_frame_time_not_set():
+    """
+    Frame time is needed, so make sure an exception gets raised.
+    """
+    nints, ngroups, nrows, ncols = 1, 5, 103, 102
+    dims = (nints, ngroups, nrows, ncols)
+    rnoise, gain = 10, 1
+    group_time, frame_time = 1.0, 1
+
+    ramp_data, gain2d, rnoise2d = setup_inputs(
+        dims, gain, rnoise, group_time, frame_time
+    )
+    ramp_data.data[0:, 0:12, :] = np.nan
+
+    ramp_data.frame_time = None  # Must be set
+
+    save_opt, algo, ncores = False, "GLS", "none"
+    with pytest.raises(UnboundLocalError):
+        slopes, cube, ols_opt, gls_opt = ramp_fit_data(
+            ramp_data, 512, save_opt, rnoise2d, gain2d, algo,
+            "optimal", ncores, test_dq_flags)
+
+
+def test_five_groups_two_ints_Poisson_noise_only():
+    """
+    I don't understand this test.
+    """
+    nints, ngroups, nrows, ncols = 2, 5, 103, 102
+    dims = (nints, ngroups, nrows, ncols)
+    rnoise, gain = 7, 2000
+    group_time, frame_time = 3.0, 1
+
+    ramp_data, gain2d, rnoise2d = setup_inputs(
+        dims, gain, rnoise, group_time, frame_time
+    )
+    ramp_data.data[0, 0, 50, 50] = 10.0
+    ramp_data.data[0, 1, 50, 50] = 15.0
+    ramp_data.data[0, 2, 50, 50] = 25.0
+    ramp_data.data[0, 3, 50, 50] = 33.0
+    ramp_data.data[0, 4, 50, 50] = 60.0
+    ramp_data.data[1, 0, 50, 50] = 10.0
+    ramp_data.data[1, 1, 50, 50] = 15.0
+    ramp_data.data[1, 2, 50, 50] = 25.0
+    ramp_data.data[1, 3, 50, 50] = 33.0
+    ramp_data.data[1, 4, 50, 50] = 160.0
+
+    save_opt, algo, ncores = False, "GLS", "none"
+    slopes, cube, ols_opt, gls_opt = ramp_fit_data(
+        ramp_data, 512, save_opt, rnoise2d, gain2d, algo,
+        "optimal", ncores, test_dq_flags)
+
+    out_slope = slopes[0][50, 50]
+    deltaDN1 = 50
+    deltaDN2 = 150
+    check = (deltaDN1 + deltaDN2) / 2.0
+    print(f"\nout_slope = {out_slope}, check = {check}")
+
+    # These arguments don't make sense to me.
+    # np.testing.assert_allclose(actual, desired, rtol, atol, ...)
+    # out_slope=slopes[0][50, 50] = 5.782827
+    # check = (deltaDN1 + deltaDN2)/2.0 = 100.0
+    # WTF?
+    # np.testing.assert_allclose(out_slope, (deltaDN1 + deltaDN2)/2.0, 75.0, 1e-6)
+    np.testing.assert_allclose(out_slope, check, 75.0, 1e-6)
+
+
+@pytest.mark.skip(reason="GLS returns NaN's, but should return zeros.")
```

Comment:
Do we expect that the problem of returning NaN's is or will be fixed in the new version of the code?

Reply:
The code as it runs right now needs a lot of work.  It runs as it did before the refactor, but the GLS code before the refactor was not complete.  The GLS code still needs a lot of work to get it to completion.  The skipped tests are a guide for many of the things that need to be fixed.